Update usr.bin.ricochet-apparmor

epidemics-scepticism opened the pull request:
Improvments/fixes to the apparmor profile

needs access to /sys/devices.../config for some graphics cards (i915 it seems, in my case), access to /proc/$PID/environ to read envvars (e.g. TOR_CONTROL_PORT), access to Tor's Control Cookie file to authenticate. Strangely also /dev/dri/ to draw the UX (not covered by <abstractions/X>?) Updated with `aa-genprof`, tested under Tails 3.0.1 and Debian Stretch desktop.

infinity0 commented:
@special The author asked me to include this in Debian and I plan to do so, unless someone tells me this is horrible - I know almost nothing about apparmor. I also asked him to add an extra commit installing the profile by default into `/etc/apparmor.d` used by Debian Ubuntu Gentoo Arch (Fedora does not have apparmor yet) so hopefully you should see an update to this PR along those lines soon as well.

infinity0 commented on contrib/usr.bin.ricochet-apparmor:
```diff
@@ -1,57 +1,41 @@
   /sys/devices/system/cpu/ r,
-
-  # Allow Ricochet to read itself
+  /usr/bin/pulseaudio rix,
```
You removed all the comments, can you leave them in? At least the non-obvious ones like for pulseaudio, or perhaps just leave all of them in.

special commented on ricochet.pro:
```diff
@@ -61,6 +61,16 @@ contains(DEFINES, RICOCHET_NO_PORTABLE) {
+                apparmor_profile.path = $${APPARMORDIR}/
+        } else {
+                apparmor_profile.path = /etc/apparmor.d/
+	}
```
nit: indent

special commented on ricochet.pro:
```diff
@@ -61,6 +61,16 @@ contains(DEFINES, RICOCHET_NO_PORTABLE) {
+        !isEmpty(APPARMORDIR) {
+                apparmor_profile.path = $${APPARMORDIR}/
+        } else {
+                apparmor_profile.path = /etc/apparmor.d/
```
Does this end up creating an /etc/apparmor.d/ on systems that don't use apparmor? I think ideal logic would be to install to `/etc/apparmor.d/` if that directory already exists, or install to `$APPARMORDIR` if explicitly specified, and otherwise not install.

One way to do this is to specify commands to execute in `apparmor_profile.extras`, instead of using `.files`. You could avoid the extra `cp` above that way too.

epidemics-scepticism commented:
On the basis that we don't know if the build environment matches the intended execution environment, I decided instead of looking for an existing `/etc/apparmor.d/` to just allow users to explicitly define if they want to include it through `APPARMOR` _(and optionally `APPARMORDIR`)_ defines.

Removed use of `system()` and replaced with with `.extra`.

Also fixed the tab/space indentiation mismatch.

infinity0 commented:
I think it was OK to install apparmor config files even if the build/runtime system(s) doesn't/don't have apparmor installed, and this pattern of pre-emptively installing config files is fairly common on Debian. But this new version is OK too.

special commented:
Yeah -- it's mostly non-Debian systems where I see that as a problem. This version looks good, thanks!

infinity0 commented:
Did anyone test `apparmor_profile.extras`? The command is getting run but the file is not getting installed. I'll hack around it for now in the Debian package but you probably want to fix it for other distros.

infinity0 commented:
Also:
~~~~
audit: type=1400 audit(XXX): apparmor="DENIED" operation="open" profile="/usr/bin/ricochet" name="/etc/xdg/QtProject/qtlogging.ini" pid=12197 comm="ricochet" requested_mask="r" denied_mask="r" fsuid=1000 ouid=0
audit: type=1400 audit(XXX): apparmor="DENIED" operation="open" profile="/usr/bin/ricochet" name="/dev/tty" pid=12197 comm="ricochet" requested_mask="r" denied_mask="r" fsuid=1000 ouid=0
audit: type=1400 audit(XXX): apparmor="DENIED" operation="open" profile="/usr/bin/ricochet" name="/usr/share/qt5/translations/qt_en.qm" pid=12197 comm="ricochet" requested_mask="r" denied_mask="r" fsuid=1000 ouid=0
audit: type=1400 audit(XXX): apparmor="DENIED" operation="mkdir" profile="/usr/bin/ricochet" name="/home/infinity0/.cache/Ricochet/" pid=12197 comm="QQmlThread" requested_mask="c" denied_mask="c" fsuid=1000 ouid=1000
audit: type=1400 audit(XXX): apparmor="DENIED" operation="open" profile="/usr/bin/ricochet" name="/usr/share/qt5/translations/qtquickcontrols_en.qm" pid=12197 comm="ricochet" requested_mask="r" denied_mask="r" fsuid=1000 ouid=0
audit: type=1400 audit(XXX): apparmor="DENIED" operation="mkdir" profile="/usr/bin/ricochet" name="/home/infinity0/.cache/Ricochet/" pid=12197 comm="QQmlThread" requested_mask="c" denied_mask="c" fsuid=1000 ouid=1000
audit: type=1400 audit(XXX): apparmor="DENIED" operation="mkdir" profile="/usr/bin/ricochet" name="/home/infinity0/.cache/Ricochet/" pid=12197 comm="QQmlThread" requested_mask="c" denied_mask="c" fsuid=1000 ouid=1000
audit: type=1400 audit(XXX): apparmor="DENIED" operation="mkdir" profile="/usr/bin/ricochet" name="/home/infinity0/.cache/Ricochet/" pid=12197 comm="QQmlThread" requested_mask="c" denied_mask="c" fsuid=1000 ouid=1000
audit: type=1400 audit(XXX): apparmor="DENIED" operation="mkdir" profile="/usr/bin/ricochet" name="/home/infinity0/.cache/Ricochet/" pid=12197 comm="QQmlThread" requested_mask="c" denied_mask="c" fsuid=1000 ouid=1000
audit: type=1400 audit(XXX): apparmor="DENIED" operation="mkdir" profile="/usr/bin/ricochet" name="/home/infinity0/.cache/Ricochet/" pid=12197 comm="QQmlThread" requested_mask="c" denied_mask="c" fsuid=1000 ouid=1000
~~~~

epidemics-scepticism commented:
Regarding the apparmor logs, can you give me an idea of what kind of system you're running that under? I don't get any kind of similar under the default Debian desktop (stretch), or Tails.

This is the results of me running it under Tails, there are a few denies but during this run it generated and published a new key, added a contact and exchanged messages.

```
Oct 23 20:26:19 amnesia kernel: audit: type=1400 audit(1508790379.543:58): apparmor="DENIED" operation="open" profile="/usr/bin/ricochet" name="/home/amnesia/.config/ibus/bus/dfea607e95134f1a9d7e30514320470e-unix-1" pid=17663 comm="ricochet" requested_mask="r" denied_mask="r" fsuid=1000 ouid=1000
Oct 23 20:26:19 amnesia kernel: audit: type=1400 audit(1508790379.544:59): apparmor="DENIED" operation="open" profile="/usr/bin/ricochet" name="/home/amnesia/.config/ibus/bus/dfea607e95134f1a9d7e30514320470e-unix-1" pid=17663 comm="ricochet" requested_mask="r" denied_mask="r" fsuid=1000 ouid=1000
Oct 23 20:26:19 amnesia kernel: audit: type=1400 audit(1508790379.676:60): apparmor="DENIED" operation="open" profile="/usr/bin/ricochet" name="/dev/tty" pid=17663 comm="ricochet" requested_mask="r" denied_mask="r" fsuid=1000 ouid=0
Oct 23 20:26:19 amnesia kernel: audit: type=1400 audit(1508790379.685:61): apparmor="DENIED" operation="open" profile="/usr/bin/ricochet" name="/etc/os-release" pid=17663 comm="ricochet" requested_mask="r" denied_mask="r" fsuid=1000 ouid=0
Oct 23 20:26:19 amnesia kernel: audit: type=1400 audit(1508790379.686:62): apparmor="DENIED" operation="open" profile="/usr/bin/ricochet" name="/etc/debian_version" pid=17663 comm="ricochet" requested_mask="r" denied_mask="r" fsuid=1000 ouid=0
```

Regarding the `.extras` for apparmor, it should work but it looks like the [debian build](https://sources.debian.net/src/ricochet-im/1.1.4-1/debian/rules/#L28) isn't defining `APPARMOR` in the `DEFINES` yet, which is required for it to create/install the apparmor profile, since special was concerned about making superflous folders for other distros that don't use apparmor (pretty much anything that's not debian-alike).

Update: I can see some of these apparmor problems, how about we kill off the section about allowing access to using the GTK themes and instead just use the pre-existing abstractions for xdg-desktop, gnome, wayland? kde is already there. There aren't any abstractions for the QT5 paths though, I will need to write an explicit exception for them.

infinity0 commented:
I'm running under a Debian testing system with some Debian unstable packages installed. I don't know enough about apparmor or ricochet to comment on your proposals, they seem reasonable but I'll let @special comment.

You linked to an old version, [this](https://anonscm.debian.org/cgit/pkg-privacy/packages/ricochet-im.git/tree/debian/rules?id=debian/1.1.4-2) is the latest released version. I do define `APPARMOR` and the `cp` command (from `.extra`) is even getting run, but it seems like this makes qmake ignore what's in `.files` - or something along those lines. For example in the [build log](https://buildd.debian.org/status/fetch.php?pkg=ricochet-im&arch=amd64&ver=1.1.4-2&stamp=1508452934&raw=0) you can see:

~~~~
cp -f /<<PKGBUILDDIR>>/contrib/usr.bin.ricochet-apparmor /<<PKGBUILDDIR>>/contrib/usr.bin.ricochet
~~~~
